### PR TITLE
[X86, MC] Recognize OSIZE=64b when EVEX.W = 1, EVEX.pp = 01

### PR DESCRIPTION
**Pull request by FreddyLeaf**
In the legacy space, if both the 66 prefix and REX.W=1 are present, the REX.W=1 takes precedence and makes OSIZE=64b.  EVEX map 4 inherits this convention, with EVEX.pp=01 and EVEX.W playing the roles of the 66 prefix and REX.W.  So if EVEX.pp=00, the OSIZE can only be 64b or 32b, depending on whether EVEX.W=1 or not.  But if EVEX.pp=01, then OSIZE is either 64b or 16b depending on whether EVEX.W=1 or not.

### CODE REVIEW

File: llvm/test/MC/Disassembler/X86/apx/imul.txt [hunk 1]
```diff
@@ -21,6 +21,11 @@
 # INTEL: {evex}	imul	r9, r9, 123
 0x62,0x54,0xfc,0x08,0x6b,0xc9,0x7b
 
+# iOSize=64 when EVEX.W = 1 and EVEX.pp = 01
```
Comment (KanRobert) [inline llvm/test/MC/Disassembler/X86/apx/imul.txt:24]: Add a separate test for this and check more instructions like llvm/test/MC/Disassembler/X86/apx/reverse-encoding.txt

Comment (FreddyLeaf) [reply llvm/test/MC/Disassembler/X86/apx/imul.txt]: [1e5d7d7](https://github.com/llvm/llvm-project/pull/103816/commits/1e5d7d7646246a394e27f642fa7df01b0c6f254b)

File: llvm/test/MC/Disassembler/X86/apx/evex-w-opsize.txt [hunk 2]
```diff
@@ -0,0 +1,74 @@
+# RUN: llvm-mc -triple x86_64 -disassemble %s | FileCheck %s --check-prefix=ATT
+# RUN: llvm-mc -triple x86_64 -disassemble -output-asm-variant=1 %s | FileCheck %s --check-prefix=INTEL
+
+# This test is to check below inherit:
+# IC_EVEX_W       -->    IC_EVEX_W_OPSIZE
+# IC_EVEX_W_B     -->    IC_EVEX_W_OPSIZE_B
+
```
Comment (KanRobert) [inline llvm/test/MC/Disassembler/X86/apx/evex-w-opsize.txt:7]: Add tests for NF and NF_ND?

Comment (KanRobert) [reply llvm/test/MC/Disassembler/X86/apx/evex-w-opsize.txt:7]: We still need to add tests for them since we accept different byte sequence in decoder.

Comment (FreddyLeaf) [reply llvm/test/MC/Disassembler/X86/apx/evex-w-opsize.txt:7]: I'll add. pls let me know if you want to do that in another commit.

Comment (KanRobert) [reply llvm/test/MC/Disassembler/X86/apx/evex-w-opsize.txt:7]: It should be added in this commit b/c it satisfy the statement `EVEX.W = 1 and EVEX.pp = 01`

Comment (FreddyLeaf) [reply llvm/test/MC/Disassembler/X86/apx/evex-w-opsize.txt:7]: 7adaff0

File: llvm/test/MC/Disassembler/X86/apx/evex-w-opsize.txt [hunk 3]
```diff
@@ -0,0 +1,74 @@
+# RUN: llvm-mc -triple x86_64 -disassemble %s | FileCheck %s --check-prefix=ATT
+# RUN: llvm-mc -triple x86_64 -disassemble -output-asm-variant=1 %s | FileCheck %s --check-prefix=INTEL
+
+# This test is to check below inherit:
+# IC_EVEX_W       -->    IC_EVEX_W_OPSIZE
+# IC_EVEX_W_B     -->    IC_EVEX_W_OPSIZE_B
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/test/MC/Disassembler/X86/apx/evex-w-opsize.txt]: [fc5a88b](https://github.com/llvm/llvm-project/pull/103816/commits/fc5a88b81c3fb0d964aadbe9922fd59b60015f30)

File: llvm/test/MC/Disassembler/X86/apx/evex-w-opsize.txt [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+## This test is to check OSIZE=64b when EVEX.W=1 and EVEX.pp = 01.
+
+## adc
+# ATT:   {evex}  adcq    $123, %r9
+# INTEL: {evex}  adc     r9, 123
+0x62,0xd4,0xfd,0x08,0x83,0xd1,0x7b
+
+# ATT:   adcq    $123, %r9, %r10
+# INTEL: adc     r10, r9, 123
+0x62,0xd4,0xac,0x18,0x83,0xd1,0x7b
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/test/MC/Disassembler/X86/apx/evex-w-opsize.txt]: 69367eb

**Review by KanRobert — APPROVED** [review]
LGTM

**Pull request merged**